ADR/ADR036-automate-patch-version-bumps.md

```diff
@@ -0,0 +1,42 @@
+# ADR036: Automate patch version bumps
+
+Date: 2024-10-31 :ghost:
+
+## Status
+
+Proposed
+
+## Context
+
+GOV.UK Forms uses [Dependabot](https://github.com/dependabot) to keep dependacies up to date. Dependancy update pull requests are raised automatically by Dependabot but a human has to approve and merge the pull requests. This can be an issue because:
+- we regularly have a large number of automated pull requests which require our attention. This means unnecessary toil, usually for the person on support
+- security patches can go unnoticed for longer than we would like and they are only applied when we take action
+
+A large number of PRs for a small team like that of GOV.UK Forms is not sustainable alongside other work, and being lax with
+dependency updates and their security fixes is not something we should be doing.
+
+### Semantice versioning
+In general, the packages consumed by GOV.UK Forms conform to [semantic versioning](https://semver.org/) rules[2], which
+allows us to make some assumptions about the changes introduced in a version change. In a semantically versioned package,
+the patch version "MUST be incremented if only backward compatible bug fixes are introduced". By definition, we can
+believe that if tests are passing after a patch version bump, then the change of version has had no negative impact.
+
+Therefore, we can safely merge a PR for a patch version change once all of its status checks have passed.
+
+Minor versions are similarly defined in the semantic versioning spec ("MUST be incremented if new, backward compatible
+functionality is introduced to the public API."), however experience tells us that package authors more frequently
+disrespect the specification for minor versions than patch versions, and as as such minor and major version bumps are
+out of scope for this ADR. A future ADR may address them.
+
+### Risks
+See the [Justification section in GOV.UK's rfc-167-auto-patch-dependencies](https://github.com/alphagov/govuk-rfcs/blob/main/rfc-167-auto-patch-dependencies.md#justification)
```

Comment:
This linked RFC is really useful context and I agree that the same applies to us in terms of it not presenting an increased risk to auto-merge based on the justifications GOV.UK lay out.

As one of their conditions for enabling auto-merging, they state that repositories must have adequate security scanning. Would it be worth outlining here what security scanning we use and that enabling this should be a prerequisite of enabling auto-merging of dependabot PRs on our repositories?

Reply:
It's also I think worth noting that as part of the conditions for automatic patching they include "[there is] sufficient test coverage". I'm not even sure what our test coverage is 😅 So I'm not confident that we have the 95% coverage that they recommend [[1](https://github.com/alphagov/govuk-rfcs/blob/main/rfc-167-auto-patch-dependencies.md#sufficient-test-coverage)].

Reply:
rspec/simplecov claim the following coverage:

- forms-admin: 96.54%
- forms-api: 96.78%
- forms-runner: 97.72%
- forms-product-page: 98.67%

I'm not sure I fully trust these though

Reply:
Thanks David! I thought that it was already really high - well done team!

Reply:
I spoke to the Lead Dev on GOV.UK about how it's been going since they enabled auto-merge of Dependabot PRs for _all_ versions iirc (excepting only on Rails mega versions like 7 -> 8) and they've had no incidents related to it which is a good indication it's working fine.

See https://github.com/alphagov/govuk-dependabot-merger

Reply:
Thanks all, I've added some clarification about the risks we discussed. 

Let me know if I've missed anything, got it wrong or you spot anything new

Reply:
I'm resolving this conversation, but feel free to reopen if needed